### ISSUE TITLE
pyenv: python3: command not found 

### ISSUE THREAD

**Issue by bollig** [aws/aws-parallelcluster#1820]
**Environment:**
 - aws-parallelcluster 2.7.0
 - Configuration file : 

```
[aws]
aws_region_name = us-east-1

[global]
cluster_template = default
update_check = false
sanity_check = true

[cluster default]
key_name = AWS-HPC-Quickstart-1YsrofaW
base_os = centos7
scheduler = slurm
master_instance_type = m5n.large
compute_instance_type = c5n.2xlarge
master_root_volume_size = 200
compute_root_volume_size = 80
vpc_settings = public
ebs_settings = myebs
cluster_type = ondemand
placement_group = DYNAMIC
placement = compute
initial_queue_size = 0
max_queue_size = 10 
disable_hyperthreading = true
cw_log_settings = cw-logs
s3_read_resource = arn:aws:s3:::*
post_install = s3://ebbollig-pcluster-userdata-XXXX/notears/notearshpc-orig.post.sh
post_install_args = "/shared/spack-0.13 /opt/slurm/log sacct.log"
dcv_settings = dcv
#enable_efa = compute

[cw_log cw-logs]
enable = false

[vpc public]
vpc_id = vpc-XXX
master_subnet_id = subnet-XXX
compute_subnet_id = subnet-XXX
use_public_ips = false

[dcv dcv]
enable = master

[ebs myebs]
shared_dir = /shared
volume_type = gp2
volume_size = 2000

[aliases]
ssh = ssh {CFN_USER}@{MASTER_IP} {ARGS}
```

**Bug description and how to reproduce:**
On a freshly booted CentOS cluster, the pyenv shims for python3 do not seem to be configured properly. 

`which python3 && python3`

![image](https://user-images.githubusercontent.com/187202/83554565-a06ff280-a4d2-11ea-97a1-50c43c88222a.png)

**Comment by bollig** [comment]
FWIW, was able to link 3.6.9 into pyenv with: 

```
$> sudo su -
#> pyenv global 3.6.9
```

Note that `sudo pyenv global 3.6.9` fails. 
![image](https://user-images.githubusercontent.com/187202/83556373-605e3f00-a4d5-11ea-9bf5-3aaa2792510b.png)

**Comment by tilne** [comment]
Hi, Evan. Thanks for reporting this issue.

This is expected behavior. Unfortunately, what you point out here is the tradeoff for using pyenv. Some of the author's thoughts on the matter can be found in [here](https://github.com/pyenv/pyenv/issues/1112#issuecomment-372509476).

That said, I think you can accomplish what you're after by running the following:
```
pyenv which python3 && python3
```

Does that help?

**Comment by bollig** [comment]
Thanks @tilne. Your suggestion is good insight, but doesn't quite solve the problem. Unfortunately in this case, the file /usr/local/pyenv/version does not exist, so the python3 shim from pyenv throws an error when running `pyenv which python3`. 

![image](https://user-images.githubusercontent.com/187202/83567192-d6b76d00-a4e6-11ea-8632-d57e03b6557a.png)

I think parallelcluster should create the version file. Alternatively, if pyenv is only intended for parallelcluster cookbooks and not to be used by end-users, then I'd suggest pyenv be removed from the end-user path (i.e., remove /etc/profile.d/pyenv.sh). 

Cheers, 
-E

**Comment by tilne** [comment]
What's the context in which you're using `pyenv which python3 && python3`? Doesn't that command throwing an error have the desired result of signaling that there's no python3 available?

Regarding your point about not exposing a pyenv configuration, I believe the reason we do that is because we currently support an operating system where the system python is too old to support an essential dependency. The pyenv system installation allows us to workaround that.

That said, we've discussed internally moving away from that eventually due in part to the issues you've identified. Because of that, I'm going to mark this as a feature request.

**Comment by bollig** [comment]
My $0.02 on "Doesn't that command throwing an error have the desired result of signaling that there's no python3 available?":

Not really. Pyenv states that python3 _is_ available from multiple python versions (e.g., 3.6.9, cookbook_virtualenv, etc.), so the error message reads along the lines of "The python3 command is not _configured_ properly. With multiple versions installed and available, end-users need to choose one to complete their configuration.", which is a totally different mindset from "python3 is not available". If the pyenv shim for python3 did not exist, a `which python3` at login would indicate that I need to install the python3 RPM, Miniconda, etc. Instead, the shim has a valid python3 command on the user-path, which must mean it is intended for end-users to use. I'd rather see a fully-configured and usable python3 on login---or no python3 command at all---instead of a partially configured version that is not intended to be used by end-users. 

Thanks for adding this as a feature request! Small gotchas like these burn hours of time. At the very least, this thread should educate users to `sudo yum install -y python3` as one way to resolve the issue.

**Comment by tilne** [comment]
> Instead, the shim has a valid python3 command on the user-path, which must mean it is intended for end-users to use.

This is a limitation of pyenv. It's not really a valid shim, even though it masquerades as one. This is definitely a gotcha.

> I'd rather see a fully-configured and usable python3 on login---or no python3 command at all---instead of a partially configured version that is not intended to be used by end-users.

The thought is to not force users or system software to use our pyenv environments (except where absolutely needed). Setting the system-wide default to be one of the interpreters or virtuialenvs we use would run the risk of breaking stuff that intended to use the default system version.

Agreeed there are several things about this approach that are not ideal. We'll discuss internally the best way to address this and update this accordingly.

**Comment by tilne** [comment]
This is addressed by #634. The solution we decided to go with was to avoid writing the script to /etc/profile.d/pyenv.sh that exposes the pyenv installation used by the ParallelCluster node daemons and cookbook code.
